File: src/Enterprise/Java/FactoryMethodExample.java
interface FactoryMethodDatabase {
    void connect();
    void query();
}

final class FactoryMethodPostgres implements FactoryMethodDatabase {
    public void connect() { System.out.println("PostgreSQL connect"); }
    public void query() { System.out.println("PostgreSQL query"); }
}

final class FactoryMethodMySql implements FactoryMethodDatabase {
    public void connect() { System.out.println("MySQL connect"); }
    public void query() { System.out.println("MySQL query"); }
}

abstract class FactoryMethodCreator {
    protected abstract FactoryMethodDatabase createDatabase();

    public final void useDatabase() {
        FactoryMethodDatabase database = createDatabase();
        database.connect();
        database.query();
    }
}

final class FactoryMethodPostgresCreator extends FactoryMethodCreator {
    protected FactoryMethodDatabase createDatabase() { return new FactoryMethodPostgres(); }
}

final class FactoryMethodMySqlCreator extends FactoryMethodCreator {
    protected FactoryMethodDatabase createDatabase() { return new FactoryMethodMySql(); }
}

public final class FactoryMethodExample {
    private FactoryMethodExample() { }

    public static void main(String[] args) {
        new FactoryMethodPostgresCreator().useDatabase();
        new FactoryMethodMySqlCreator().useDatabase();
    }
}

File: src/Enterprise/Kotlin/FactoryMethodExample.kt
private interface FactoryMethodDatabase {
    fun connect()
    fun query()
}

private class FactoryMethodPostgres : FactoryMethodDatabase {
    override fun connect() = println("PostgreSQL connect")
    override fun query() = println("PostgreSQL query")
}

private class FactoryMethodMySql : FactoryMethodDatabase {
    override fun connect() = println("MySQL connect")
    override fun query() = println("MySQL query")
}

private abstract class FactoryMethodCreator {
    protected abstract fun createDatabase(): FactoryMethodDatabase

    fun useDatabase() {
        val database = createDatabase()
        database.connect()
        database.query()
    }
}

private class FactoryMethodPostgresCreator : FactoryMethodCreator() {
    override fun createDatabase(): FactoryMethodDatabase = FactoryMethodPostgres()
}

private class FactoryMethodMySqlCreator : FactoryMethodCreator() {
    override fun createDatabase(): FactoryMethodDatabase = FactoryMethodMySql()
}

fun main() {
    FactoryMethodPostgresCreator().useDatabase()
    FactoryMethodMySqlCreator().useDatabase()
}
